pyresample/gradient/__init__.py

```diff
@@ -406,3 +440,194 @@ def _concatenate_chunks(chunks):
+    return wrapper
+
+
+class RBGradientSearchResampler(BaseResampler):
+    """Resample using gradient search based bilinear interpolation."""
+
+    def __init__(self, source_geo_def, target_geo_def):
+        """Init GradientResampler."""
+        super().__init__(source_geo_def, target_geo_def)
+        logger.debug("/!\\ Instantiating an experimental GradientSearch resampler /!\\")
+        self.indices_xy = None
+
+    def precompute(self, **kwargs):
+        """Precompute resampling parameters."""
+        from pyresample.resampler import resample_blocks
+        if self.indices_xy is None:
+            self.indices_xy = resample_blocks(gradient_resampler_indices, self.source_geo_def, [], self.target_geo_def,
+                                              chunks=(2, CHUNK_SIZE, CHUNK_SIZE), dtype=float)
+
+    @ensure_data_array
+    def compute(self, data, method="bilinear", cache_id=None, **kwargs):
+        """Perform the resampling."""
+        from pyresample.resampler import resample_blocks
+
+        if method == "bilinear":
+            fun = block_bilinear_interpolator
+        elif method in ["nearest_neighbour", "nn"]:
+            fun = block_nn_interpolator
+        else:
+            raise ValueError(f"Unrecognized interpolation method {method} for gradient resampling.")
+
+        chunks = list(data.shape[:-2]) + [CHUNK_SIZE, CHUNK_SIZE]
+
+        res = resample_blocks(fun, self.source_geo_def, [data.data], self.target_geo_def,
+                              dst_arrays=[self.indices_xy],
+                              chunks=chunks, dtype=data.dtype, **kwargs)
+
+        coords = _fill_in_coords(self.target_geo_def, data.coords, data.dims)
+
+        res = xr.DataArray(res, attrs=data.attrs.copy(), dims=data.dims, coords=coords)
+        res.attrs["area"] = self.target_geo_def
+        return res
+
+
+def ensure_3d_data(func):
+    """Ensure the data is in three dimensions."""
+    @wraps(func)
+    def wrapper(data, *args, **kwargs):
+        """Wrap around the original function."""
+        if data.ndim == 2:
+            data_3d = data[np.newaxis, :, :]
+        else:
+            data_3d = data
+
+        resampled = func(data_3d, *args, **kwargs)
+
+        if data.ndim == 2:
+            resampled = resampled.squeeze(0)
+        return resampled
+
+    wrapper.__doc__ += "\n\nThe input data can be 2d, or 3d with the two last axes being respectively `y` and `x`."
+    return wrapper
+
+
+@ensure_3d_data
+def gradient_resampler(data, source_area, target_area, method='bilinear'):
+    """Do the gradient search resampling."""
+    from pyproj.transformer import Transformer
+
+    try:
+        src_x, src_y = source_area.get_proj_coords()
+        transformer = Transformer.from_crs(target_area.crs, source_area.crs)
+    except AttributeError:
+        src_x, src_y = source_area
+        # TODO: this is bad. We don't want to perform gradient search in lat/lon space (poles, dateshift line)
```

Comment:
This seems like a big TODO item. What's the state of this?

Reply:
This part of the code cannot resample to or from Swath. I removed the todo and raise NotImplementedErrors now instead.
Resampling to and from swath will be done in a following PR, as I'd like to keep the one for area defs only.